File: integ-tests/python/tests/test_abort_handlers.py
import pytest
import asyncio
import time
from baml_client import b
from baml_py import AbortController
from baml_py.errors import BamlAbortError


@pytest.mark.asyncio
async def test_manual_cancellation():
    """Test manual abort of a function call"""
    abort_controller = AbortController()

    async def abort_after_delay():
        await asyncio.sleep(0.1)
        abort_controller.abort()

    task = asyncio.create_task(
        b.FnFailRetryExponentialDelay(
            retries=5,
            initial_delay_ms=100,
            baml_options={"abort_controller": abort_controller},
        )
    )
    asyncio.create_task(abort_after_delay())

    with pytest.raises(Exception) as exc_info:
        await task

    assert (
        "abort" in str(exc_info.value).lower()
        or "cancel" in str(exc_info.value).lower()
    )


@pytest.mark.asyncio
async def test_streaming_cancellation():
    """Test abort of a streaming operation"""
    abort_controller = AbortController()

    stream = b.stream.TestAbortFallbackChain(
        input="test streaming", baml_options={"abort_controller": abort_controller}
    )

    async def abort_after_delay():
        await asyncio.sleep(0.05)
        abort_controller.abort()

    asyncio.create_task(abort_after_delay())

    values = []
    cancelled = False
    try:
        async for value in stream:
            values.append(value)
            # If we've collected some values and controller is aborted, should stop soon
            if abort_controller.aborted and len(values) > 5:
                break
    except Exception:
        cancelled = True
        # Expected to be cancelled

    # Either it was cancelled or it stopped early after abort
    # The test succeeds if stream was interrupted (by exception or early stop)
    assert cancelled or abort_controller.aborted
    # Delete the stream object
    del stream


def test_sync_cancellation():
    """Test abort in synchronous context"""
    from baml_client.sync_client import b as sync_b

    abort_controller = AbortController()

    def abort_after_delay():
        time.sleep(0.05)  # Reduced delay to abort faster
        abort_controller.abort()

    import threading

    threading.Thread(target=abort_after_delay).start()

    # Since the function might complete quickly, we'll check if it was aborted
    # or if an exception was raised
    try:
        sync_b.FnFailRetryConstantDelay(
            retries=5, delay_ms=100, baml_options={"abort_controller": abort_controller}
        )
        # If we got here, check that the controller was at least triggered
        assert abort_controller.aborted, "Function completed but abort wasn't triggered"
    except Exception as e:
        # This is expected - either aborted or some other error
        assert (
            "abort" in str(e).lower()
            or "cancel" in str(e).lower()
            or abort_controller.aborted
        )


def test_streaming_sync_cancel_after_ten_messages():
    """Cancel a synchronous stream after receiving 10 messages"""
    from baml_client.sync_client import b as sync_b

    abort_controller = AbortController()

    stream = sync_b.stream.PromptTestStreaming(
        input="Programming languages are fun to create",
        baml_options={"abort_controller": abort_controller},
    )

    msgs = []
    raised_exception = False
    try:
        for msg in stream:
            print(f"msg: {msg}")
            msgs.append(str(msg))
            if len(msgs) >= 10:
                abort_controller.abort()
                break
        stream.get_final_response()

    except Exception as e:
        # It's acceptable if aborting mid-stream raises
        print(f"Exception: {e}")
        raised_exception = True

    print("stopped streaming")

    # In normal runs we expect to have aborted after >=10 messages
    # But in CI without API keys, requests may error early; accept exceptions too
    assert abort_controller.aborted or raised_exception
    if abort_controller.aborted:
        assert len(msgs) >= 10, "Expected at least 10 streamed messages before aborting"
    del stream


@pytest.mark.asyncio
async def test_streaming_async_cancel_after_ten_messages():
    """Cancel an async stream after receiving 10 messages"""
    abort_controller = AbortController()

    stream = b.stream.PromptTestStreaming(
        input="Programming languages are fun to create",
        baml_options={"abort_controller": abort_controller},
    )

    msgs = []
    raised_exception = False
    try:
        async for msg in stream:
            print(f"msg: {msg}")
            msgs.append(str(msg))
            if len(msgs) >= 10:
                abort_controller.abort()
                break
    except Exception as e:
        # It's acceptable if aborting mid-stream raises or network errors occur
        print(f"Exception: {e}")
        raised_exception = True

    assert abort_controller.aborted or raised_exception
    if abort_controller.aborted:
        assert len(msgs) >= 10, "Expected at least 10 streamed messages before aborting"
    del stream


@pytest.mark.asyncio
async def test_early_abort():
    """Test that already-aborted controller prevents execution"""
    abort_controller = AbortController()
    abort_controller.abort()

    with pytest.raises(BamlAbortError):
        await b.ExtractName(
            text="John Doe", baml_options={"abort_controller": abort_controller}
        )


def test_baml_abort_error_import_and_instanceof():
    try:
        raise BamlAbortError("Operation was aborted")
    except Exception as e:
        assert isinstance(e, BamlAbortError)


@pytest.mark.asyncio
async def test_normal_operation():
    """Test that operations work normally without abort controller"""
    result = await b.ExtractName(text="My name is Alice")
    assert isinstance(result, str)
    assert "alice" in result.lower()


@pytest.mark.asyncio
async def test_multiple_aborts():
    """Test multiple concurrent aborted operations"""
    tasks = []

    for i in range(10):
        controller = AbortController()
        task = asyncio.create_task(
            b.FnFailRetryConstantDelay(
                retries=3, delay_ms=50, baml_options={"abort_controller": controller}
            )
        )
        tasks.append(task)

        # Abort at random times
        asyncio.create_task(abort_after(controller, 0.01 * (i + 1)))

    results = await asyncio.gather(*tasks, return_exceptions=True)

    # All should have raised exceptions
    for result in results:
        assert isinstance(result, Exception)


async def abort_after(controller, delay):
    """Helper to abort after a delay"""
    await asyncio.sleep(delay)
    controller.abort()


@pytest.mark.asyncio
async def test_abort_timing():
    """Test that abort happens quickly"""
    abort_controller = AbortController()
    start_time = time.time()

    # Cancel after 250ms
    asyncio.create_task(abort_after(abort_controller, 0.25))

    with pytest.raises(Exception):
        await b.FnFailRetryExponentialDelay(
            retries=5,
            initial_delay_ms=100,
            baml_options={"abort_controller": abort_controller},
        )

    elapsed = time.time() - start_time
    # Should abort within ~300ms (250ms delay + processing)
    assert elapsed < 0.4, f"Took too long to abort: {elapsed}s"


@pytest.mark.asyncio
async def test_abort_time_out():
    """Test that abort happens quickly"""
    abort_controller = AbortController(timeout_ms=250)
    start_time = time.time()

    with pytest.raises(Exception):
        await b.FnFailRetryExponentialDelay(
            retries=5,
            initial_delay_ms=100,
            baml_options={"abort_controller": abort_controller},
        )

    elapsed = time.time() - start_time
    # Should abort within ~300ms (250ms delay + processing)
    assert elapsed < 0.4, f"Took too long to abort: {elapsed}s"

File: integ-tests/rust/tests/test_abort_handlers.rs
//! Abort handler tests - ported from test_abort_handlers_test.go
//!
//! Tests for abort/cancellation functionality including:
//! - Manual cancellation
//! - Timeout cancellation
//! - Streaming cancellation
//! - Retry chain cancellation
//! - Fallback chain cancellation

use baml::CancellationToken;
use rust::baml_client::sync_client::B;
use std::thread;
use std::time::{Duration, Instant};

/// Test manual cancellation - Go: TestAbortHandlerManualCancellation
#[test]
fn test_manual_cancellation() {
    let token = CancellationToken::new();
    let token_clone = token.clone();

    token.cancel();

    // This should be cancelled before completion
    let result = B.TestOpenAI.with_cancellation_token(Some(token));

    // Cancel after 100ms in another thread
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(10));
        token_clone.cancel();
    });
    let result = result.call("write a short story about a cat (100 words)");

    // Go: assert.Error(t, err) && assert.Contains(t, err.Error(), "context canceled")
    assert!(result.is_err(), "Expected error from cancellation");
    let error = result.unwrap_err();
    let error_str = format!("{:?}", error).to_lowercase();
    assert!(
        error_str.contains("cancel") || error_str.contains("abort"),
        "Expected cancellation error, got: {}",
        error_str
    );
}

/// Test timeout cancellation - Go: TestAbortHandlerTimeoutCancellation
#[test]
fn test_timeout_cancellation() {
    // Create token with 200ms timeout
    let token = CancellationToken::new_with_timeout(Duration::from_millis(200));

    // This should timeout before all retries complete
    let result = B
        .TestRetryExponential
        .with_cancellation_token(Some(token))
        .call();

    // Go: assert.Error(t, err) && assert.Contains(t, err.Error(), "deadline exceeded")
    assert!(result.is_err(), "Expected error from timeout");
    let error = result.unwrap_err();
    let error_str = format!("{:?}", error).to_lowercase();
    assert!(
        error_str.contains("cancel")
            || error_str.contains("timeout")
            || error_str.contains("deadline")
            || error_str.contains("aborterror"),
        "Expected timeout/cancellation error, got: {}",
        error_str
    );
}

/// Test streaming cancellation - Go: TestAbortHandlerStreamingCancellation
#[test]
fn test_streaming_cancellation() {
    let token = CancellationToken::new();
    let token_clone = token.clone();

    let stream = B
        .TestFallbackClient
        .with_cancellation_token(Some(token))
        .stream();

    assert!(stream.is_ok(), "Expected successful stream creation");
    let mut stream = stream.unwrap();

    // Cancel after 50ms in another thread
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(50));
        token_clone.cancel();
    });

    let mut count = 0;
    for partial in stream.partials() {
        if partial.is_ok() {
            count += 1;
        }
    }

    // Go: assert.Less(t, count, 10, "Stream should have been cancelled early")
    // Stream should have stopped early due to cancellation
    assert!(
        count < 10,
        "Stream should have been cancelled early, got {} partials",
        count
    );
}

/// Test retry chain cancellation - Go: TestAbortHandlerRetryChainCancellation
#[test]
fn test_retry_chain_cancellation() {
    // guarantee all resources are initialized ahead of time, so we don't measure construction time
    rust::baml_client::init();

    // Create token with 300ms timeout
    let token = CancellationToken::new_with_timeout(Duration::from_millis(300));

    let result = B.TestRetryExponential.with_cancellation_token(Some(token));

    let start = Instant::now();
    let result = result.call();
    let duration = start.elapsed();

    // Go: assert.Error(t, err)
    assert!(result.is_err(), "Expected error from cancellation");

    // Go: assert.Less(t, duration, 400*time.Millisecond, "Should have cancelled before all retries")
    // Should have been cancelled before all exponential retries complete
    // Exponential delays would sum up to more than 300ms
    assert!(
        duration < Duration::from_millis(400),
        "Should have cancelled before all retries, took {:?}",
        duration
    );
}

/// Test fallback chain cancellation - Go: TestAbortHandlerFallbackChainCancellation
#[test]
fn test_fallback_chain_cancellation() {
    // guarantee all resources are initialized ahead of time, so we don't measure construction time
    rust::baml_client::init();

    // Create token with 150ms timeout
    let token = CancellationToken::new_with_timeout(Duration::from_millis(150));

    let start = Instant::now();
    let result = B
        .TestFallbackClient
        .with_cancellation_token(Some(token))
        .call();
    let duration = start.elapsed();

    // Go: assert.Error(t, err)
    assert!(result.is_err(), "Expected error from cancellation");

    // Go: assert.Less(t, duration, 200*time.Millisecond, "Should have cancelled during fallback chain")
    assert!(
        duration < Duration::from_millis(200),
        "Should have cancelled during fallback chain, took {:?}",
        duration
    );
}

/// Test no interference with normal operation - Go: TestAbortHandlerNoInterferenceWithNormalOperation
#[test]
fn test_no_interference_normal_operation() {
    // Test that operations complete normally when not cancelled
    let result = B.ExtractNames.call("My name is John Doe");

    // Should complete successfully (or fail due to LLM, but not due to cancellation)
    match result {
        Ok(names) => {
            assert!(!names.is_empty(), "Expected non-empty result");
        }
        Err(e) => {
            let error_str = format!("{:?}", e).to_lowercase();
            // Go: We're just checking that cancellation doesn't interfere when not triggered
            assert!(
                !error_str.contains("cancel") && !error_str.contains("deadline"),
                "Should not have cancellation error, got: {}",
                error_str
            );
        }
    }
}

/// Test multiple concurrent cancellations - Go: TestAbortHandlerMultipleConcurrentCancellations
#[test]
fn test_multiple_concurrent_cancellations() {
    let token = CancellationToken::new();
    let token_for_cancel = token.clone();

    let handles: Vec<_> = (0..3)
        .map(|_| {
            let t = token.clone();
            thread::spawn(move || B.TestRetryConstant.with_cancellation_token(Some(t)).call())
        })
        .collect();

    // Cancel all operations after 100ms
    thread::spawn(move || {
        thread::sleep(Duration::from_millis(100));
        token_for_cancel.cancel();
    });

    // Collect all results
    let mut error_count = 0;
    for handle in handles {
        let result = handle.join().expect("Thread panicked");
        if result.is_err() {
            error_count += 1;
        }
    }

    // Go: All operations should have been cancelled
    // When aborted, you should get the same error as the context
    assert_eq!(error_count, 3, "Expected all 3 operations to be cancelled");
}
